### PR TITLE
Add sidebar to `revisions#show`

### PR DESCRIPTION
**Pull request by MattBudz**
### Summary

This PR add's a sidebar to the `revisions#show` view for easier navigation.
The sidebar is loaded via ajax to not hinder the performance of the view.

### Check List

- [x] Added a CHANGELOG entry

### CODE REVIEW

File: app/controllers/revisions_controller.rb [hunk 1]
```diff
@@ -52,28 +60,29 @@ def load_node
...
-                Issue.find(params[:issue_id])
-              elsif params[:note_id]
-                @node.notes.find(params[:note_id])
-              else
-                raise 'Unable to identify record type'
-              end
+  def set_variables_from_params
+    if params[:card_id]
+      load_list
+      @path = sidebar_project_board_list_card_revisions_path
```
Comment (aapomm) [inline app/controllers/revisions_controller.rb:66]: I don't think we need an extra "sidebar" route for every type of item. A single one will suffice  like: `sidebar_project_revisions_path(params)`. That way, we can pass the same `@path` value for each type.

File: config/routes.rb [hunk 2]
```diff
@@ -32,7 +32,11 @@
         member { post :move }
         resources :cards, except: [:index] do
           member { post :move }
-          resources :revisions, only: [:index, :show]
+          resources :revisions, only: [:index, :show] do
+            collection do
```
Comment (aapomm) [inline config/routes.rb:36]: See comment above about using a single route vs a sidebar route for every type of item.

File: app/views/revisions/sidebar.html.erb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<% if @issue %>
```
Comment (aapomm) [inline app/views/revisions/sidebar.html.erb:1]: I imagine we have the `@record` here. If so, we can probably simplify this file to:

```
<%= render "#{@record.class.name}/sidebar" %>
```

Comment (MattBudz) [reply app/views/revisions/sidebar.html.erb:1]: That won't work because we need to render `nodes/sidebar` for notes and evidence

File: app/controllers/revisions_controller.rb [hunk 4]
```diff
@@ -54,28 +62,37 @@ def load_node
...
-              elsif params[:note_id]
-                @node.notes.find(params[:note_id])
-              else
-                raise 'Unable to identify record type'
-              end
-  rescue ActiveRecord::RecordNotFound
-    flash[:error] = 'Record not found'
-    redirect_to :back
+  def set_revision_params
+    @revision_params = { :"#{@record.class.name.downcase}_id" => @record.id }
```
Comment (aapomm) [inline app/controllers/revisions_controller.rb:66]:
```suggestion
    @revision_params = { "#{@record.class.name.downcase}_id" => @record.id }
```

not sure if we need this to be a symbol vs a string

File: app/controllers/revisions_controller.rb [hunk 5]
```diff
@@ -54,28 +62,37 @@ def load_node
...
+    elsif params[:note_id]
+      load_node
+      @record = @node.notes.find(params[:note_id])
+    else
+      raise 'Unable to identify record type'
     end
+
+    set_revision_params
+
+  rescue ActiveRecord::RecordNotFound
```
Comment (aapomm) [inline app/controllers/revisions_controller.rb:94]: I'm not fond of using errors as control flow and we may be better off using if else here instead. Instead of using `find` we can use `find_by_id` so that it returns nil if the record is missing instead of throwing an error. Then, we can check for the existence of `@record` here instead.

Comment (MattBudz) [reply app/controllers/revisions_controller.rb]: I removed this rescue block completely since there is no valid case for a record to not be found.

File: spec/features/revisions_spec.rb [hunk 6]
```diff
@@ -6,7 +6,7 @@
   describe 'when the record has 2 revisions' do
     let(:record) do
       with_versioning do
-        create(:issue)
+        create(:issue, node_id: current_project.issue_library.id)
```
Comment (etdsoft) [inline spec/features/revisions_spec.rb:9]: Why is this needed? The Issues factory already accounts for the right Node association.

File: config/routes.rb [hunk 7]
```diff
@@ -51,7 +51,7 @@
 
       resources :evidence, concerns: :multiple_destroy, controller: 'issues/evidence', only: [:index, :new]
       resources :nodes, only: [:show], controller: 'issues/nodes'
-      resources :revisions, only: [:index, :show]
+      resources :revisions, only: [:index, :show]
```
Comment (etdsoft) [inline config/routes.rb:54]: extra space here, L78, L82

@aapomm this could have been a peer-review suggestion?

File: app/controllers/revisions_controller.rb [hunk 8]
```diff
@@ -54,28 +62,33 @@ def load_node
...
+      @revision_params[:board_id] = @record.board.id
+      @revision_params[:list_id] = @record.list.id
+    end
   end
 
-  def load_sidebar
-    if params[:evidence_id] || params[:note_id]
-      load_node
-    elsif params[:card_id]
+  def set_variables_from_params
```
Comment (etdsoft) [inline app/controllers/revisions_controller.rb:75]: this is the wrong name for this helper. It's setting the `@record` variable yes (so `set_record`, but it's also loading collections (e.g. `load_list`, `load_issues`).

It's doing too much. What about extracting the loading of Sidebar collections (which you don't need any more until the ajax call to a controller concern as well?

You duplicate two `if else if` trees, but at least it's easier to follow.

File: app/controllers/revisions_controller.rb [hunk 9]
```diff
@@ -54,28 +62,33 @@ def load_node
...
+      load_issues
+      @record = @issue
+    elsif params[:note_id]
+      load_node
+      @record = @node.notes.find(params[:note_id])
+    else
+      raise 'Unable to identify record type'
     end
+
+    set_revision_params
```
Comment (etdsoft) [inline app/controllers/revisions_controller.rb:92]: more over, we're also calling another helper! Should this last one be called from a before_action or the action itself?

There's definitely too much going on here.

File: app/views/revisions/show.html.erb [hunk 10]
```diff
@@ -1,7 +1,18 @@
...
+  <%= content_tag(
+    :div,
+    content_tag(
+      :div,
+      spinner_tag,
+      class: 'd-flex align-items-center justify-content-center h-100'
+    ),
+    data: {
+      behavior: 'fetch',
+      path: project_revisions_sidebar_path(@revision_params)
```
Comment (etdsoft) [inline app/views/revisions/show.html.erb:13]: once extracted (see below) this will have to be a resource-dependent route, so probably set in the controller (for easier logic)

File: app/views/layouts/nested_node_resource.html.erb [hunk 11]
```diff
@@ -1,5 +1,7 @@
 <% content_for :content do %>
-  <%= render "nodes/sidebar" %>
+  <% content_for :sidebar do %>
```
Comment (etdsoft) [inline app/views/layouts/nested_node_resource.html.erb:2]: does `content_for` need to be nested inside the other `content_for`? Seems a bit untidy
